Long range fork resolution

akoptelov opened the pull request:
I've added long range fork handling. Also I've introduced an intermediate state for fork range detection.

I'm going to collect block pairs for different forks (both long and short ranges) and add them along with tests.

Re: #5 
Fix: #31

akoptelov commented:
@binier I've migrated long forks to new version, please review.

binier commented on snarker/src/consensus/consensus_reducer.rs:
```diff
@@ -87,6 +117,40 @@ impl ConsensusState {
+                    time: meta.time(),
+                    compared_with: tip_hash,
+                    decision: if take {
+                        ConsensusLongRangeForkDecision::Take(why)
```
For consistency, can we use `Use` and `Ignore` instead of `Take` and `Keep`?

binier commented on snarker/src/consensus/consensus_state.rs:
```diff
@@ -37,6 +37,29 @@ impl ConsensusShortRangeForkDecision {
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum ConsensusLongRangeForkDecision {
+    Keep(ConsensusLongRangeForkResolutionKind),
```
Can we rename this for consistency?
`Keep` -> `Ignore`
`Take` -> `Use`

akoptelov replied:
Keep/take is a terminology from Mina implementation, like here:
https://github.com/MinaProtocol/mina/blob/472c82cded7362594be3681cba5d41862477d817/src/lib/consensus/proof_of_stake.ml#L2959

I thought that using it is a good idea.

akoptelov replied:
I agree about consistency. We can choose either Mina terminology or Ignore/Use.

Also I would get rid of negated enum for the short range fork.

binier replied:
Cool. Let's use mina terminology then for short range fork also.

akoptelov commented:
@binier I've synced up all the names between long and short forks, also extracted short range decision to a function and added some tests for short range forks.

binier reviewed (approved):
Looks good. Feel free to merge.
